### ISSUE TITLE
[Bug] dashboardService.getCustomerDashboards displays wrong parameters for it to work

### ISSUE THREAD

**Issue by X1folta** [thingsboard/thingsboard#9590]
**Describe the bug**

According to the [dashboard.service.ts](https://github.com/thingsboard/thingsboard/blob/8c25e00c2f7950aab6ffc891096daa436b9c52eb/ui-ngx/src/app/core/http/dashboard.service.ts#L65) it says its:

```
public getCustomerDashboards(customerId: string, pageLink: PageLink, config?: RequestConfig): Observable<PageData<DashboardInfo>> {
    return this.http.get<PageData<DashboardInfo>>(`/api/customer/${customerId}/dashboards${pageLink.toQuery()}`,
      defaultHttpOptionsFromConfig(config));
  }
```

Therefore p1 customerId, p2 pageLink and optional p3 being a config.

But console.logging dashboardService reveals its parameters are:

` getCustomerDashboards: function getCustomerDashboards(f, y, C, s)`

So a 4th parameter that is missing in the info-window about getCustomerDashboards. That fourth parameter is a boolean for 

&includeCustomers=_boolean_

**Our Thingsboardversion:**

Thingsboard Version PE 3.6

**Comment by devaskim** [comment]
You provided link for CE version. Unfortunately there is no public link to PE widget's API. So only `console.log()` may provide more info about `dashboardService` object and its methods.

**Comment by X1folta** [comment]
@devaskim Thanks for the info. I suspected that. Nevertheless the autocompletion inside the PE environment suggests the same amount and description of arguments as described and leads to the function's error.

**Comment by devaskim** [comment]
Unfortunately autocompletion is incomplete as well both in CE and PE

**Comment by devaskim** [comment]
Seems like this issue has the lowest priority in TB task tracker, because such topic has been raised several time before.

**Comment by X1folta** [comment]
Thats okay. If one day the mystical "ThingsBoard Marketplace" really should become reality then everything that facilitates the entry into creating widgets must receive a higher priority.

**Comment by devaskim** [comment]
It would be historical day)))

**Comment by devaskim** [comment]
Just compare [API docs](https://thingsboard.cloud/swagger-ui/#/dashboard-controller/getCustomerDashboardsUsingGET) and the output of `console.log(dashboardService.getCustomerDashboards)` in browser DevTools Console

```
const pageObj = self.ctx.pageLink(pageSize, pageNumber /* start from 0 */, searchTextString /* optional, null by default */, sortOrder /* optional, null by default, values: 'ASC' or 'DESC');
const customerId = 'YOUR_CUSTOMER_UUID';
const includeCustomers = true; // or false
dashboardService.getCustomerDashboards(customerId, pageObj, includeCustomers).subscribe(
    response => console.log('Response', response),
    error=> console.log('Error', error)
);
```

**Comment by merrittholmes7** [comment]
Further to my previous message . There is a clear difference between PE and CE versions of the dashboardService. 

CE definition (taken from demo.thingsboard.io):

getCustomerDashboards(h, r, p) {
                return this.http.get(/api/customer/${h}/dashboards${r.toQuery()}, (0,
                e.T7)(p))
            }

PE definition (taken from analytics.hyscore.co)

getCustomerDashboards(f, y, C, s) {
 let _ = /api/customer/${y}/dashboards${C.toQuery()};
 return f && (_ += "&includeCustomers=true"),
 this.http.get(_, (0,
 t.T7)(s))
 }

So please advise how to use the PE version. What should ‘f' be. I assume ‘s’ in the PE defintion is the same as 'p’ in the CE one.

**Comment by merrittholmes7** [comment]
I should ad that your example above erro

> Try to compare [API docs](https://thingsboard.cloud/swagger-ui/#/dashboard-controller/getCustomerDashboardsUsingGET) and the output of `console.log(dashboardService.getCustomerDashboards)` in browser DevTools Console
> 
> ```
> const pageObj = self.ctx.pageLink(pageSize, pageNumber /* start from 0 */, searchTextString /* optional, null by default */, sortOrder /* optional, null by default, values: 'ASC' or 'DESC');
> const customerId = 'YOUR_CUSTOMER_UUID';
> const includeCustomers = true; // or false
> dashboardService.getCustomerDashboards(customerId, pageObj, includeCustomers).subscribe(
>     response => console.log('Response', response),
>     error=> console.log('Error', error)
> );
> ```
> 
> The output of `console.log(dashboardService.getCustomerDashboards)` ![Скриншот 08-03-2024 171029](https://private-user-images.githubusercontent.com/86909129/311228476-f48b8f3c-8d08-42e1-8783-96f7823032ae.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk5MDE4MjEsIm5iZiI6MTcwOTkwMTUyMSwicGF0aCI6Ii84NjkwOTEyOS8zMTEyMjg0NzYtZjQ4YjhmM2MtOGQwOC00MmUxLTg3ODMtOTZmNzgyMzAzMmFlLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzA4VDEyMzg0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVhM2ZmNjc3YmZiMDRlZTIyMDExYTE0MGI1MTFjYWFiZWI5YzljMzQ5OTgyNWYzYjljM2FlNDkxYTc0YzI0ZTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.O5PI3OF0-n4rDJPMoQuOvSOZVJYKhK3ca0-0GrjMTlA)

I should add that your example above errors in my PE environment with Cannot read properties of undefined (reading 'pageLink')

This is because for some reason in the PE environment, the pageLink is the 3rd variable but its unclear what the first is meant to be

**Comment by devaskim** [comment]
> So please advise how to use the PE version

My example code was for PE, because original question is about PE version of the API.

> the pageLink is the 3rd variable but its unclear what the first is meant to be

Did you try just swap 1st and 3rd parameters from my code example.
```
dashboardService.getCustomerDashboards(includeCustomers, customerId, pageObj).subscribe(.......);
```

**Comment by merrittholmes7** [comment]
Your example code was for CE not PE as it failed in my PE environment. I have now worked out the issues here. So here are 2 examples for people who are struggling with this also:

getCustomerDashboards:
CE:
dashboardService.getCustomerDashboards(CustomerId, pageLink, optional config)
PE:
dashboardService.getCustomerDashboards(includeCustomers (this is mandatory), CustomerId, pageLink, optional config)
getUserDashboards:
CE:
doesn't exist
PE:
dashboardService.getUserDashboards(userId,operation (I use null as this is mandatory in PE), pageLink, optional config)

In essence though, the Bug here is that the autocomplete details in PE are not correct as they display the CE APIs and unfortunately we do not have access to the codebase for PE so every user has to dig around trying to find the answer in the code that does not normal variable names.

Please update the PE version to include the PE API automcomplete details.

Thanks

**Comment by devaskim** [comment]
> Please update the PE version to include the PE API automcomplete details.

Sad, but you are just another one who is suffering from autocomplete issues. We have been asking TB team to update autocomplete for 1-3 years - no changes yet.

> Your example code was for CE not PE as it failed in my PE environment

Explain me how both my examples might be for CE considering that CE version doesn't have nested customer at all (I mean `includeCustomers ` option in both my examples).

Seems like you didn't read my second anwer
![Скриншот 08-03-2024 190023](https://github.com/thingsboard/thingsboard/assets/86909129/66e1eb49-2f22-4bdd-8a3d-e497ea07a065)

**Comment by merrittholmes7** [comment]
Yes you are right, the example was for PE but the includeCustomer boolean was put in the wrong place and I read it like the CE version which has the optional config parameter. Apologies there.

this is the issue though in that we are just having to reverse engineer all of the methods because the autocomplete is the only docs we have to go on and its for the wrong version. I have requested a bug be raised for this too.

Thanks for your support. Its appreciated.

**Comment by devaskim** [comment]
@merrittholmes7 As you are PE user, (if you didn't yet) open the bug on [Thingsboard Portal (Service Desk)](https://thingsboard-portal.atlassian.net/servicedesk/customer/portals). Hope you will get more info there or maybe even estimation how long we should wait for the full coverage of autocompletion.

**Comment by mykhailokornieiev** [comment]
We're letting you know that this issue, labeled as "help wanted", has been open for a while. This label is used to seek support from the GitHub community, especially for issues that fall outside the scope of the ThingsBoard support team's expertise. Unfortunately, there hasn't been any community input or progress on this issue over the month(s). We're closing this ticket to keep our repository well-maintained and focused on active issues. We appreciate your patience and understanding.
